raiden/transfer/mediated_transfer/initiator_manager.py

```diff
@@ -64,129 +68,183 @@ def events_for_cancel_current_route(transfer_description) -> List[Event]:
     return [unlock_failed]
 
 
-def cancel_current_route(payment_state: InitiatorPaymentState) -> List[Event]:
+def cancel_current_route(
+        payment_state: InitiatorPaymentState,
+        initiator_state: InitiatorPaymentState,
+) -> List[Event]:
     """ Cancel current route.
 
     This allows a new route to be tried.
     """
-    assert can_cancel(payment_state), 'Cannot cancel a route after the secret is revealed'
+    assert can_cancel(initiator_state), 'Cannot cancel a route after the secret is revealed'
 
-    transfer_description = payment_state.initiator.transfer_description
+    transfer_description = initiator_state.transfer_description
 
-    payment_state.cancelled_channels.append(payment_state.initiator.channel_identifier)
-    payment_state.initiator = None
+    payment_state.cancelled_channels.append(initiator_state.channel_identifier)
 
     return events_for_cancel_current_route(transfer_description)
 
 
-def handle_block(
+def maybe_try_new_route(
         payment_state: InitiatorPaymentState,
-        state_change: Block,
+        initiator_state: InitiatorTransferState,
+        transfer_description: TransferDescriptionWithSecretState,
+        available_routes: List[RouteState],
+        channelidentifiers_to_channels: ChannelMap,
+        pseudo_random_generator: random.Random,
+        block_number: BlockNumber,
+) -> TransitionResult:
+    events: List[Event] = list()
+    if can_cancel(initiator_state):
+        cancel_events = cancel_current_route(payment_state, initiator_state)
+
+        sub_iteration = initiator.try_new_route(
+            old_initiator_state=initiator_state,
+            channelidentifiers_to_channels=channelidentifiers_to_channels,
+            available_routes=available_routes,
+            transfer_description=transfer_description,
+            pseudo_random_generator=pseudo_random_generator,
+            block_number=block_number,
+        )
+
+        events.extend(cancel_events)
+        events.extend(sub_iteration.events)
+        assert sub_iteration.new_state
```

Comment:
`try_new_route` may fail, instead of asserting here the case where there is no good route should be handled.

Reply:
Depends on what you mean by "fail". If we don't find a new route, we'll just return the current `payment_state`, otherwise a new state will be created and appended to the list of initiator transfers.
So the failure here is safe in a way that nothing will happen until the lock for that transfer has expired.

Reply:
Hummm, I see, the `old_initiator_state` is being passed as an argument to `try_new_route`. This doesn't make sense to me. IMO transfers should be independent of one another, the `initiator` should return `None`, and here, the error should be checked.

Reply:
I agree, this was resolved.